### PR TITLE
Implement onToken option for parse

### LINKED ISSUE #205: [Feature Request] A way to get the AST and tokens in one go
Currently there seems to be no way to get the AST and tokens without tokenizing the same source code twice (once with the TokenizerStream and once with the parse function).
This is wasteful, especially for bigger source files.
I propose to either:
1. Add an `onToken` callback (similar to `onComment` except it is called for all tokens)
2. Add a boolean option `tokens` which adds the tokens to the result of the parse function

### PR DESCRIPTION
**Pull request by nzakas**
This pull request introduces a new feature to handle tokens during parsing, along with corresponding documentation updates and tests. The most important changes include adding a new `onToken` option, updating the parser to support this option, and adding tests to ensure the new functionality works correctly.

I took a cue from the design of `onComment` and so `onToken` receives three arguments:

1. `type` - the string name of the token
2. `value` - the representation of the token in the source text
3. `loc` - the location of the token

Other decisions made:

* `onToken()` is not called for white space or comment tokens. I'm assuming anyone who cares about comments will use `onComment` and that no one really cares about whitespace tokens.

Fixes #205

### New Feature: Token Handling

* [`docs/parsing.md`](diffhunk://#diff-f6518598c49353c6dd23850c3af87f8213a478ec071ba5b8feba4d28570beba2R30): Added documentation for the new `onToken` option, which allows a handler to be called for every non-whitespace, non-comment token in the parsing source. [[1]](diffhunk://#diff-f6518598c49353c6dd23850c3af87f8213a478ec071ba5b8feba4d28570beba2R30) [[2]](diffhunk://#diff-f6518598c49353c6dd23850c3af87f8213a478ec071ba5b8feba4d28570beba2R107-R113)

### Parser Updates

* [`lib/parser/create.js`](diffhunk://#diff-ce97140e9ee81523283d0d32734140c7fb92f3169d57f27ee5d11fa1b1af6d2fL318-R347): Updated the `createParser` function to include the `onToken` option and modified the token processing logic to call the `onToken` handler when appropriate.

### Tests

* [`lib/__tests/parse.js`](diffhunk://#diff-c4e30655c2d8cb0f22ee50b0e16a062fa38a8631331b8acef0c67c2588b17f33R374-R443): Added new test cases to verify the `onToken` functionality, both with and without location information.

### CODE REVIEW

**Comment by nzakas** [comment]
ping @lahmatiy

**Comment by lahmatiy** [comment]
On one hand, I don't fully support the proposed changes. The `onComment` approach was introduced as a temporary solution while we explore better ways to handle comments. Continuing down this path isn't ideal.

On the other hand, I recognize that some users

**Comment by lahmatiy** [comment]
On one hand, I don't fully support the proposed changes. The `onComment` approach was introduced as a temporary solution while we explore better ways to handle comments. Continuing to extend this approach feels suboptimal.

On the other hand, I recognize that some users expect expect access to tokens while working with a parser, as some parsers offer this functionality (despite tokenization being a separate concern from parsing). Since this is an opt-in feature and could provide significant utility, I think it's reasonable to include an option for token access.

### Feedback on the proposed implementation

As a user, I  would expect the `onToken()` callback to be invoked for every token,, including whitespaces and comments, regardless of the existence of a dedicated `onComment()` callback. This would allow users to collect all tokens and reconstruct the original source or portion of it. The implementation should prioritize performance and memory efficiency, even at the expense of convenience. This aligns with the initial request for the feature (#205). Otherwise, the implementation offers little advantage over accessing tokens directly via `TokenStream()` (which is used internally by the parser):

```js
import { TokenStream, tokenize } from 'css-tree';

const tokenStream = new TokenStream(css, tokenize);
tokenStream.forEachToken((type, start, end, index) => {
  // Process tokens as needed
});
```

Or even more simply, using `tokenize()` directly:

```js
import { tokenize } from 'css-tree';

tokenize(css, (type, start, end) => {
  // Process tokens as needed
});
```

The parser extends the functionality of `TokenStream`, which uses the `tokenize()` function to retrieve basic token information (`type`, `start`, `end`). The `TokenStream` builds additional structures for parsing operations like lookahead, skipping tokens. It also computes line and column positions (using `OffsetToLocation`) and determines balanced token pairs.

Key `TokenStream` methods relevant to this PR:

- `forEachToken()`: A low-cost method to iterate over all tokens. It avoids creating objects, slicing source strings, or decoding token names, returning only the `type`, `start`, `end`, and index of each token. Additional details can be derived from this:
  - Token type name:
    ```js
    import { tokenNames } from 'css-tree';

    console.log(tokenNames[tokenType]); // Map token type code to name
    ```
    Using token type codes directly is more performant; token names are mainly for debugging.
    
  - Token "value":
    ```js
    const tokenValue = source.substring(start, end);
    ```
    Note: Token values are often unnecessary for many token types (e.g., single characters like `:`, `,`, or fixed strings like `CDO`). For others, users may need to trim the value depending on context (e.g., at-rules, hashes, functions). This variability makes it best to leave value extraction to the user.
- `dump()`: Designed for debugging, this method constructs an object for each token using `forEachToken()`. It decodes token types, generates token values as substrings, and provides balance token indices. It lacks `start` and `end` fields but these can be reconstructed during iteration.

### Proposal for onToken()

To ensure performance and universality, the `onToken()` callback should directly utilize `forEachToken()` and receive `type`, `start`, and `end`. Additional enhancements to `forEachToken()`, such as access to balance tokens or location computation (i.e. `loc`), could be considered.

For users prioritizing convenience over performance, a `tokens: boolean` option could be introduced. When enabled, it would return an array of token objects similar to those in `dump()` but with additional fields (e.g., `start` and `end`). However, this introduces a challenge: the parser currently returns a single AST. Several options to address this:

1. Introducing a breaking change in a new major version to return an object, e.g., `{ ast, tokens }`.
2. Returning an object only when `tokens: true` is specified.
3. Adding a new method, such as `parseExtended()` (need to come up with a better name), that accepts additional options and returns `{ ast, ... }`.

I personally prefer the last option.

---

In summary, the `onToken()` callback should prioritize performance and be tightly integrated with `forEachToken()`. Additional options for retrieving all tokens as an array could be provided for those prioritizing convenience over efficiency. This would balance usability and performance without redundant token generation.

File: lib/parser/create.js [hunk 1]
```diff
@@ -315,21 +315,35 @@ export function createParser(config) {
...
+                    }
+                    return;
+                }
+
+                if (type === WhiteSpace) {
+                    return;
+                }
+
+                if (onToken) {
                     const loc = parser.getLocation(start, end);
```
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/parser/create.js:343]: Can you explain more about this? Are you calculating the location from scratch every time `parser.getLocation()` is called?

In my JSON parser ([Momoa](https://github.com/humanwhocodes/momoa)), the lines and columns are tracked inline with tokenization so there's no additional calculation. [Acorn](https://github.com/acornjs/acorn/) similarly tracks line/column position along with source offset when you pass `locations` as an option to `parse()`. Could we maybe do something similar if `onToken` is present?

From a syntax analysis perspective, the full location is always necessary, so I'd prefer to have that produced automatically rather than adding another (potentially expensive) step after the fact.

**Comment by nzakas** [comment]
My perspective working on other parsers: `onToken` satisfies all of the use cases for static analysis. I don't think there's a need to build up and return the tokens array as part of the result of `parse()`

(In ESLint we use Acorn under the hood, which has a similar `onToken` callback.)

File: lib/parser/create.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (lahmatiy) [reply lib/parser/create.js:343]: Lines and columns for the source are computed lazily on the first call to [`getLocation()`](https://github.com/csstree/csstree/blob/76fb90ae0422b9f9bc45bbe008778a7db7ddb213/lib/parser/create.js#L252-L262), using [OffsetToLocation](https://github.com/csstree/csstree/blob/master/lib/tokenizer/OffsetToLocation.js) provided `parse()` is called `positions: true`. The location object has the following structure:

```ts
{
    source: string | null,
    start: {
        offset: number,
        line: number,
        column: number
    },
    end: {
        offset: number,
        line: number,
        column: number
    }
}
```

Such structure allocates at least 106 bytes of memory per token, plus the computational overhead of object creation, values read, and math. This cost applies to every token, even if their location data is never used.

In most cases, token locations remain unused. For example, in your case, locations are only needed when an error or warning is encountered. Computing location data for all tokens in advance would result in unnecessary memory consumption and processing time.

**Comment by lahmatiy** [comment]
I reviewed Acorn's `onToken` approach, which can accept an array. In this case, Acorn populates the array with token objects. That's a 4th option, and I find it more favorable than the others.

**Comment by lahmatiy** [comment]
For clarity: when I mentioned "a 4th option, and I find it more favorable than the others" I was referring to the option of allowing `onToken` to accept either a callback or an array. This would enable flexibility, users could provide a callback for token processing or supply an array to have it populated with token objects.

File: lib/parser/create.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nzakas) [reply lib/parser/create.js:343]: Right, I totally understand that reasoning. FWIW, ESLint's location structure is the same (minus `source`). My suggestion is that when someone provides `onToken`, they are explicitly saying they want to pay that cost in order to get the token information.

Separately from that, it looks like location [calculation](https://github.com/csstree/csstree/blob/0afee3d88916a2860b48144537027cfb2f48c080/lib/tokenizer/OffsetToLocation.js#L11) always runs on the entire input from start to finish to get all location information. I think that may be part of the performance penalty as you're definitely computing values you might not need.

In Momoa, for example, I have a [`CharCodeReader` class](https://github.com/humanwhocodes/momoa/blob/main/js/src/char-code-reader.js) that is used to read char codes from the input. As it does so, it automatically tracks the line and column such that you can call `locate()` at any point in time and retrieve the offset, line, and column for that character. I'm wondering if something like that would reduce the cost of calculating locations here.

**Comment by nzakas** [comment]
I've updated this PR:

1. `onToken` can now be either a function or array
2. `onToken` receives type, start, end (all numbers)
3. `onToken` produces every token including comments and whitespace
4. `onToken` does not return location information

File: lib/__tests/parse.js [hunk 2]
```diff
@@ -371,6 +372,71 @@ describe('parse', () => {
...
+
+        it('as function', () => {
+            const actual = [];
+            parse(source, {
+                onToken(type, start, end) {
+                    actual.push({ type, start, end });
+                }
+            });
+
+            assert.deepStrictEqual(actual, [
```
Comment (lahmatiy) [inline lib/__tests/parse.js:386]: Nit: The array can be stored as `expected` right after `const source = '...'` to reduce repetition, since it's the same for both test cases.

File: lib/parser/create.js [hunk 3]
```diff
@@ -315,21 +315,29 @@ export function createParser(config) {
...
 
-        const { context = 'default', onComment } = options;
+        const { context = 'default', onComment, onToken } = options;
 
         if (context in parser.context === false) {
             throw new Error('Unknown context `' + context + '`');
         }
 
-        if (typeof onComment === 'function') {
+        if (typeof onComment === 'function' || typeof onToken === 'function' || Array.isArray(onToken)) {
```
Comment (lahmatiy) [inline lib/parser/create.js:324]: Nit: `onToken` can be checked for `isArray()` before the block and turned into a function if needed. This allows to simplify things, since for the block `onToken` will be a function or undefined.

```js
if (Array.isArray(onToken)) {
    onToken = (type, start, end) => ({ type, start, end });
}
```

**Comment by lahmatiy** [comment]
@nzakas Overall, I'm ready to accept the changes. I’ve left a couple of comments, but I can handle those myself after the merge or with a commit to your PR. The only thing that needs to be fixed before merging is restoring the removed spaces in `docs/parsing.md`. If you're okay with it, I can address everything myself.

I appreciate that the API changes in this PR are minimal. While it might not be fully convenient for end users yet, we can enhance it in future iterations. I'm considering how to include all the necessary functionality in `onToken()` callback calls and similar. My current preference is to provide an object API within the callback (as `this` or 4th argument), offering a minimal set of useful methods, such as retrieving location, token value (a substring from source), block start/end, and so on.

File: lib/parser/create.js [hunk 4]
```diff
@@ -315,21 +315,29 @@ export function createParser(config) {
...
+                            ? source.slice(start + 2, end - 2)
+                            : source.slice(start + 2, end);
+
+                        onComment(value, loc);
+                    }
+                }
 
-                    onComment(value, loc);
+                if (typeof onToken === 'function') {
+                    onToken(type, start, end);
```
Comment (lahmatiy) [inline lib/parser/create.js:338]: I completely forgot! We need a 4th parameter, `index`. It's passed as the 4th argument in `forEachToken` and should be forwarded to `onToken`. This is important for determining the token's position and accessing values in other arrays like `balance`.

**Comment by lahmatiy** [comment]
I introduced a token iteration API. Using this API, the problem described in issue #301 can be addressed as follows:

```js
import { parse, tokenTypes } from 'css-tree';

const emitWarning = (text, location) => {
    console.warn(text, location);
};

parse(css, {
    onToken(type, start, end, index) {
        switch (type) {
            case tokenTypes.BadString:
                emitWarning('Bad string', this.getRangeLocation(start, end));
                break;

            case tokenTypes.BadUrl:
                emitWarning('Bad url', this.getRangeLocation(start, end));
                break;

            default:
                if (this.isBlockOpenerTokenType(type)) {
                    if (this.getBlockTokenPairIndex(index) === -1) {
                        emitWarning('No closing pair', this.getRangeLocation(start, end));
                    }
                } else if (this.isBlockCloserTokenType(type)) {
                    if (this.getBlockTokenPairIndex(index) === -1) {
                        emitWarning('No opening pair', this.getRangeLocation(start, end));
                    }
                }
        }
    }
});
```

File: docs/parsing.md [hunk 5]
```diff
@@ -98,42 +99,49 @@ csstree.parse('example { foo; bar: 1! }', {
 
 ### onComment
 
-Type: `function(value, loc)` or `null`
```
[the remark this answers is not in the document]
Comment (nzakas) [reply docs/parsing.md:101]: Hmm. Looks like VS Code is doing this automatically on save. :thinking:

**Comment by nzakas** [comment]
That looks really nice! Just fixed the whitespace issue in the docs.

**Comment by nzakas** [comment]
@lahmatiy is there anything else that needs to be done before merging this?

**Comment by nzakas** [comment]
@lahmatiy when do you expect this change to be released?
